static/app/components/charts/useChartXRangeSelection.tsx

```diff
@@ -0,0 +1,415 @@
+        clampedXMaxPixel,
+        clampedXMinPixel,
+        xMaxPixel,
+        yMinPixel,
+      });
+
+      setState({
+        selection: newSelection,
+        actionMenuPosition,
+      });
+
+      onSelectionEnd?.(newSelection, clearSelection);
+    },
+    [onSelectionEnd, clearSelection]
+  );
+
+  const handleOutsideClick = useCallback(
+    (event: MouseEvent) => {
+      let el = event.target as HTMLElement | null;
+
+      // Propagate the click event to the parent elements until we find the element that has the
+      // data-explore-chart-selection-region attribute. This is used to prevent the selection from
+      // being cleared if the user clicks within an 'inbound' region.
+      while (el) {
+        if (el.dataset?.exploreChartSelectionRegion !== undefined) {
+          return;
+        }
+        el = el.parentElement;
+      }
+
+      clearSelection();
+    },
+    [clearSelection]
+  );
+
+  useEffect(() => {
+    if (disabled || !state?.selection) return;
+
+    window.addEventListener('click', handleOutsideClick, {capture: true});
+
+    // eslint-disable-next-line consistent-return
+    return () => {
+      window.removeEventListener('click', handleOutsideClick, {capture: true});
+    };
+  }, [handleOutsideClick, disabled, state?.selection]);
+
+  const enableBrushMode = useCallback(() => {
+    const chartInstance = chartRef.current?.getEchartsInstance();
+    chartInstance?.dispatchAction({
+      type: 'takeGlobalCursor',
+      key: 'brush',
+      brushOption: CHART_X_RANGE_BRUSH_OPTION,
+    });
+  }, [chartRef]);
+
+  useEffect(() => {
+    if (disabled) {
+      return;
+    }
+
+    const chartInstance = chartRef.current?.getEchartsInstance();
+
+    // Re-draw the box in the chart when a new selection is made
+    if (state?.selection) {
+      chartInstance?.dispatchAction({
+        type: 'brush',
+        areas: [
+          {
+            brushType: 'lineX',
+            coordRange: state.selection.range,
+            coordRanges: [state.selection.range],
+            panelId: state.selection.panelId,
+          },
+        ],
+      });
+
+      // We re-connect the group after drawing the box, so that the cursor is synced across all charts again.
+      // Check the onBrushStart handler for more details.
+      if (chartsGroupName) {
+        echarts?.connect(chartsGroupName);
+      }
+    }
+
+    // Activate brush mode on load and when we re-draw the box/clear the selection
+    enabbleBrushModeFrameRef.current = requestAnimationFrame(() => {
+      enableBrushMode();
+    });
+
+    // eslint-disable-next-line consistent-return
+    return () => {
+      if (enabbleBrushModeFrameRef.current)
+        cancelAnimationFrame(enabbleBrushModeFrameRef.current);
+      if (tooltipFrameRef.current) cancelAnimationFrame(tooltipFrameRef.current);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state, disabled, enableBrushMode, chartRef.current, chartsGroupName, deps]);
```

Comment:
Would `chartRef.current` trigger useEffect if it's a ref? Sounds like it might not.
Would it make sense to use `chartRef` instead as a dep to satisfy eslint?